src/core/search/search.cc

```diff
@@ -25,65 +27,142 @@ AstExpr ParseQuery(std::string_view query) {
+  tmp->clear();
+  tmp->reserve(matched->size() + current->size());
+
+  set_union(matched->begin(), matched->end(), current->begin(), current->end(),
+            back_inserter(*tmp));
 
-  merge(matched.begin(), matched.end(), tmp->begin(), tmp->end(), back_inserter(*current));
-  current->erase(unique(current->begin(), current->end()), current->end());
+  IndexResult result{move(*tmp)};
+  *tmp = current.TakeOwnedOrEmpty();  // If current has a backing array, keep it
+  return result;
 }
 
-// Merge results from matched with current, keep sorted.
-void MergeResults(vector<DocId>&& matched, vector<DocId>* current, vector<DocId>* tmp) {
-  swap(*current, *tmp);  // Move current to tmp so we can directly write into current
-  current->clear();
-  current->reserve(matched.size() + tmp->size());
+// Merge (AND) matched and current
+IndexResult MergeAND(IndexResult&& matched, IndexResult&& current, vector<DocId>* tmp) {
+  tmp->clear();
+  tmp->reserve(min(matched->size(), current->size()));
 
-  set_intersection(matched.begin(), matched.end(), tmp->begin(), tmp->end(),
-                   back_inserter(*current));
+  set_intersection(matched->begin(), matched->end(), current->begin(), current->end(),
+                   back_inserter(*tmp));
+
+  IndexResult result{move(*tmp)};
+  *tmp = current.TakeOwnedOrEmpty();  // If current has a backing array, keep it
+  return result;
 }
 
 struct BasicSearch {
+  using LogicOp = AstLogicalNode::LogicOp;
+
+  BasicSearch(const FieldIndices* indices) : indices_{indices}, tmp_vec_{} {
+  }
+
   // Get casted sub index by field
   template <typename T> T* GetIndex(string_view field) {
     static_assert(is_base_of_v<BaseIndex, T>);
-    auto index = indices->GetIndex(field);
+    auto index = indices_->GetIndex(field);
     DCHECK(index);  // TODO: handle not existing erorr
     auto* casted_ptr = dynamic_cast<T*>(index);
     DCHECK(casted_ptr);  // TODO: handle type errors
     return casted_ptr;
   }
 
-  vector<DocId> Search(monostate, string_view) {
-    return {};
+  // Collect all index results from F(C[i])
+  template <typename C, typename F> vector<IndexResult> GetSubResults(C&& container, F&& f) {
+    vector<IndexResult> sub_results(container.size());
+    for (size_t i = 0; i < container.size(); i++)
+      sub_results[i] = f(container[i]);
+    return sub_results;
   }
 
-  vector<DocId> Search(const AstTermNode& node, string_view active_field) {
-    // Select active indices: search in all text indices if none is selected
-    vector<TextIndex*> selected_indices;
-    if (active_field.empty())
-      selected_indices = indices->GetAllTextIndices();
-    else
-      selected_indices = {GetIndex<TextIndex>(active_field)};
+  // Efficiently unify multiple sub results with specified logical op
+  IndexResult UnifyResults(vector<IndexResult>&& sub_results, LogicOp op) {
+    if (sub_results.empty())
+      return vector<DocId>{};
+
+    auto op_func = op == AstLogicalNode::AND ? MergeAND : MergeOR;
```

Comment:
Here's an idea, take it or leave it as you wish :)

Since `MergeOR` and `MergeAND` are essentially copy pasted with the only modification being whether `set_union` or `set_intersection` is called, and here you're switching over them, maybe you could implement a single function and pass to it the operation to perform? (`set_union` or `set_intersection`)

Reply:
I'll merge them for now, with negation I'll have to split them, as each case will be handled a little differently

Reply:
No need to merge for now if that'll change so soon :)